### PR TITLE
Implement Bedrock embeddings

### LINKED ISSUE #2381: Add embeddings for Amazon Bedrock
Currently only supported in the python version of langchain.

### PR DESCRIPTION
**Pull request by cwoolum**
This PR adds the ability to use Bedrock to create embeddings. The only model in Bedrock that currently supports embeddings is Titan so the scope of the changes are limited to Titan. The end user may specify an alternative model name if future support is added for other models.

Fixes #2381 

Twitter: @chriswoolum

### CODE REVIEW

**Comment by cwoolum** [comment]
Results for integration tests

```
  console.log
    [
        0.11376953, 0.083984375,    0.21582031, -0.103515625,   1.9296875,
          0.671875,  0.31640625, -0.0012664795,  0.024414062,  0.35546875,
      0.0107421875,  0.07763672,   0.071777344,    0.2265625, -0.10546875,
       0.048828125, -0.27929688,    0.30664062,    -0.515625,  0.46679688,
         -0.390625,  0.17578125,   -0.80078125, -0.029785156,  -0.8203125,
       -0.24414062,    0.609375,     -0.765625,     0.359375, -0.19726562,
       -0.27734375,     1.15625,    0.27539062,     -1.09375,     -0.5625,
         1.0078125,  0.49804688,    -0.6796875,  0.087890625,   0.4296875,
         0.3515625,  0.08691406,      0.484375,  -0.06738281,    0.734375,
       0.123046875, 0.100097656,    -0.5234375,    0.4453125,     1.40625,
        -0.1796875,    1.171875,     0.5703125,  0.083496094, -0.06298828,
       -0.33203125,   0.7578125,   0.088378906,   0.60546875,  -0.5390625,
       0.123046875,  0.23535156,    -1.0546875,   0.95703125,   0.3671875,
       0.013671875,     -0.3125,       0.15625,   -0.7109375,       0.375,
        0.41015625,  0.42578125,     0.2265625,    1.1328125,  -0.6484375,
        0.34960938,    1.015625,     -0.734375,        0.625, 0.020751953,
        0.28515625,  0.06347656,     -0.828125,     -0.21875, -0.40234375,
        -0.6640625,    -0.96875,   -0.45703125, 0.0001296997,   -0.734375,
        0.23925781, -0.07910156,     1.3203125,     1.265625,  0.19433594,
        0.16601562, -0.66796875,   -0.43359375,   0.48828125,    -0.03125,
      ... 1436 more items
    ]

      at Object.log (src/embeddings/tests/bedrock.int.test.ts:11:11)

  console.log
    [
      [
          0.11376953, 0.083984375,    0.21582031, -0.103515625,   1.9296875,
            0.671875,  0.31640625, -0.0012664795,  0.024414062,  0.35546875,
        0.0107421875,  0.07763672,   0.071777344,    0.2265625, -0.10546875,
         0.048828125, -0.27929688,    0.30664062,    -0.515625,  0.46679688,
           -0.390625,  0.17578125,   -0.80078125, -0.029785156,  -0.8203125,
         -0.24414062,    0.609375,     -0.765625,     0.359375, -0.19726562,
         -0.27734375,     1.15625,    0.27539062,     -1.09375,     -0.5625,
           1.0078125,  0.49804688,    -0.6796875,  0.087890625,   0.4296875,
           0.3515625,  0.08691406,      0.484375,  -0.06738281,    0.734375,
         0.123046875, 0.100097656,    -0.5234375,    0.4453125,     1.40625,
          -0.1796875,    1.171875,     0.5703125,  0.083496094, -0.06298828,
         -0.33203125,   0.7578125,   0.088378906,   0.60546875,  -0.5390625,
         0.123046875,  0.23535156,    -1.0546875,   0.95703125,   0.3671875,
         0.013671875,     -0.3125,       0.15625,   -0.7109375,       0.375,
          0.41015625,  0.42578125,     0.2265625,    1.1328125,  -0.6484375,
          0.34960938,    1.015625,     -0.734375,        0.625, 0.020751953,
          0.28515625,  0.06347656,     -0.828125,     -0.21875, -0.40234375,
          -0.6640625,    -0.96875,   -0.45703125, 0.0001296997,   -0.734375,
          0.23925781, -0.07910156,     1.3203125,     1.265625,  0.19433594,
          0.16601562, -0.66796875,   -0.43359375,   0.48828125,    -0.03125,
        ... 1436 more items
      ],
      [
          -0.25976562,   -0.328125,       -0.375,     0.30273438,
            0.5234375,   0.5078125,      0.15625, -0.00064468384,
          -0.42773438,    1.046875,  -0.32421875,       1.234375,
          -0.09326172,     -0.3125, -0.037597656,    -0.18457031,
          -0.34570312, -0.78515625,    -0.890625,     0.37890625,
           0.40234375, -0.06201172,  0.087402344,     0.56640625,
          -0.25976562,  0.43359375,   0.37695312,    -0.22558594,
             -0.28125, -0.13964844,  0.027832031,        0.65625,
          -0.21679688, -0.67578125,   -0.3046875,     0.28515625,
             1.328125,   -0.515625,  -0.16992188,    -0.33984375,
            0.5234375,  0.56640625,      1.03125,     0.03540039,
             0.390625,   0.3359375,    0.2734375,    -0.08203125,
            -1.234375,  0.79296875,  -0.50390625,       0.609375,
             1.046875, -0.26757812,   0.19238281,     0.44921875,
           0.16503906,  0.66015625,          0.5,     -0.3203125,
           0.61328125, -0.51953125,    -0.328125,     -0.3515625,
          -0.42578125,  0.55078125,  0.056396484,     -0.1328125,
               -0.625,      0.4375,         0.25,        0.53125,
           0.18554688,     1.15625,   0.09472656,    -0.03100586,
              0.40625,   -0.578125,   0.55078125,     0.44921875,
          0.083984375, -0.31445312,  -0.23632812,     0.78515625,
          -0.31640625,   0.7265625,  -0.41015625,     0.14746094,
        0.00011920929,  0.10058594,  -0.16210938,     0.44921875,
              0.84375,  0.05908203,   -0.6171875,       0.671875,
          -0.42382812,   -0.671875,    0.9140625,     0.10107422,
        ... 1436 more items
      ],
      [
           0.13964844,  -0.28515625,   0.107910156,   0.072265625,    1.3046875,
          -0.22265625,  -0.17773438, -0.0002632141,     0.5546875,   -0.1796875,
          -0.43945312,   0.53515625,    0.16796875,     -0.828125,     0.640625,
          0.103515625,   0.17675781,    0.07080078,     -0.765625,    0.4609375,
           0.13964844,     0.546875,   -0.44921875,  0.0036621094,  0.068847656,
          -0.26367188,  -0.10058594,     -1.078125,    -0.7578125,  -0.51953125,
           -0.6484375,   -0.6796875,      0.765625,   -0.85546875,  -0.13183594,
             0.484375,     1.015625,   -0.20898438,  0.0009765625,      0.03125,
           0.26757812,     1.203125,      0.546875,  -0.024902344,      1.03125,
            -0.703125,  -0.08691406,      0.203125,    0.27929688,   0.87109375,
           0.29882812,   0.30078125,      0.859375,    0.45507812,   0.14257812,
           0.37695312,   0.47265625,   -0.48828125,    0.29101562,  0.017333984,
        -0.0045166016,     -1.15625,  -0.045654297,    0.90234375, -0.067871094,
          -0.08496094,  -0.61328125,    0.03955078,      0.546875,  0.006713867,
           -0.1171875,  -0.20507812,    0.47460938,    0.13476562,   0.94140625,
          0.087402344, 0.0048828125,    0.27734375,   -0.24414062,   0.41992188,
          -0.16015625,    1.0859375,    -0.1171875,    0.58203125,  -0.21484375,
             -0.21875,     0.640625,   0.056152344, 0.00012397766,  -0.66796875,
           0.22265625,   0.48632812,       0.28125,      -0.78125,  -0.37890625,
             0.109375,      -1.4375,    -1.0234375,    -0.5078125,   -0.2421875,
        ... 1436 more items
      ],
      [
            -0.3515625,  0.24023438,       0.6875,   -0.44140625,
              1.046875, -0.16503906,   0.22070312, -0.0006904602,
            0.12060547,  0.22265625,    -0.515625,  -0.037597656,
           0.057128906,  0.32421875,  0.006958008,   -0.10595703,
            -0.4140625, -0.10888672,    -0.703125,     0.2421875,
             0.3203125, -0.35351562,       -0.625,    0.72265625,
          -0.036376953,  -0.4140625,   0.22851562,   -0.34960938,
             -0.265625,  0.20898438,   0.29882812,      -0.40625,
             0.4140625,   -0.265625,    0.4609375,    0.12890625,
            0.62109375,  0.11328125,    0.7890625,  -0.022827148,
             0.7578125,  0.31835938, -0.041015625,    -1.1328125,
             0.3046875,        0.25,  -0.38476562,    0.27734375,
           0.110839844,  0.13867188,   0.38085938,   -0.24023438,
            0.33398438,    0.703125,      -0.3125, -0.0014648438,
           -0.62890625, -0.50390625,   0.16210938,     -0.109375,
           -0.26171875, -0.69140625,   0.34765625,    0.13378906,
             0.5078125,   -0.515625,   0.11767578,       -0.4375,
           0.045410156,  0.20507812,   0.34179688,   -0.67578125,
           -0.18359375, 0.053222656,  -0.04638672,    0.50390625,
          -0.063964844, -0.23046875,  -0.07910156,    0.43359375,
           -0.10449219,    0.796875,   0.16210938,    0.84765625,
           -0.32421875, -0.28320312, -0.041015625,    0.36132812,
        0.000116825104,  0.62109375,   -0.3671875,        0.4375,
             0.2109375,  -0.3359375,      0.59375,   0.052978516,
            0.08496094,    0.578125,        0.625,   -0.25390625,
        ... 1436 more items
      ],
      [
           0.33203125,    -0.296875,    -0.5859375,    0.14355469,    0.8984375,
            0.8203125,  -0.08642578, -0.0005683899,    0.37109375,  -0.29492188,
        -0.0068359375,     1.265625,   -0.32226562,   -0.18945312,     -0.84375,
          -0.19433594,      0.34375,     1.2421875,   0.032714844,  -0.29882812,
          -0.86328125,   0.10595703,       -0.6875,     0.8828125,    0.5703125,
           0.45703125, -0.052734375,  -0.022949219,      -0.34375,   -0.1640625,
           -1.9765625,    -2.984375,        -0.875,    -1.1953125,   0.16601562,
          -0.73046875,   -0.6796875,   -0.14648438,     1.3359375,    0.7734375,
             -0.65625,   -0.6640625,    0.59765625,    -0.8515625,    -0.359375,
          -0.41210938,   0.09667969,    -0.9921875,    0.18164062,   0.20507812,
           0.18164062,   0.14453125,    0.42773438,  -0.122558594,   0.08300781,
            0.3671875,    1.3671875,   -0.04345703,   -0.76953125,           -1,
           0.18554688,   -0.5234375,   -0.17871094,    -0.8203125,   -0.7578125,
           0.06933594,        0.875,    0.35546875,     -0.015625,         0.75,
             1.328125,  -0.24804688,     0.8203125,   -0.34960938, -0.053710938,
           0.23828125,  -0.30859375,     0.2421875,     0.9609375,  -0.71484375,
          -0.85546875,     0.109375,    -0.0234375,    0.29492188,   0.03564453,
          -0.33398438,  -0.31835938,   -0.93359375, 0.00011634827,   0.33984375,
          -0.06738281,   0.06640625,     -0.265625,    0.37890625,   -0.9296875,
           0.65234375,    -0.203125,   -0.60546875,   -0.30078125,   0.38476562,
        ... 1436 more items
      ],
      [
            0.515625, -0.021972656,     0.5859375,   -0.21679688,  0.26367188,
         -0.38867188,   0.07714844, -0.0011672974,      -0.71875, -0.30859375,
           0.7265625,  -0.34570312,    0.50390625,       1.03125,     -1.0625,
         -0.18457031,   0.31445312,    0.36523438,      -0.78125,  0.27734375,
            -1.03125,  -0.25195312,   -0.68359375,      1.515625,   -0.328125,
          0.30078125,   0.41992188,     -1.609375,   -0.87109375,  -1.4140625,
              0.5625,   0.37695312,    0.59765625,     -0.515625,   -0.390625,
             0.65625,  0.038085938,   -0.29101562,    0.03881836,  0.38867188,
            0.765625,   -1.3984375,       0.71875,     0.5390625, 0.010070801,
         -0.96484375,    0.7109375,    0.28320312,       1.03125,  0.49023438,
        -0.079589844,   0.21777344,    0.20117188,    0.07128906, -0.05908203,
           0.6171875,       -0.375,   0.061279297,      1.421875,   -0.734375,
        -0.076171875,    0.5390625,    -0.6484375,   -0.74609375,  0.23632812,
            1.640625,   -0.6171875,   -0.45507812,    0.34179688,        1.25,
          0.73046875,    -0.546875,    0.30859375,   -0.67578125,  0.48046875,
         0.092285156,  -0.47070312,      0.890625,  -0.107910156, 0.064453125,
          -0.5546875,  -0.13183594,   -0.41992188,     0.8671875,  -0.9140625,
            0.796875,  -0.47460938,   -0.07421875, 0.00012779236, -0.18652344,
         -0.14257812,       1.3125,      1.296875,  0.0021362305,  0.46484375,
          0.17089844, -0.037353516,    0.68359375,    -0.6171875,   -1.296875,
        ... 1436 more items
      ]
    ]

      at Object.log (src/embeddings/tests/bedrock.int.test.ts:27:11)

 PASS  src/embeddings/tests/bedrock.int.test.ts
  ✓ Test BedrockEmbeddings.embedQuery (470 ms)
  ✓ Test BedrockEmbeddings.embedDocuments (426 ms)
  ✓ Test end to end with HNSWLib (839 ms)

Test Suites: 1 passed, 1 total
Tests:       3 passed, 3 total
Snapshots:   0 total
Time:        2.766 s, estimated 3 s
```

**Comment by jacoblee93** [comment]
Nice! Looks like linting is failing if you don't mind running it?

File: langchain/package.json [hunk 1]
```diff
@@ -673,6 +673,7 @@
   "license": "MIT",
   "devDependencies": {
     "@aws-crypto/sha256-js": "^5.0.0",
+    "@aws-sdk/client-bedrock-runtime": "^3.422.0",
```
Comment (jacoblee93) [inline langchain/package.json:676]: Didn't know this was out - we should probably switch our other Bedrock integrations to use this right?

**Comment by cwoolum** [comment]
I ran the linting and it's failing on `import/no-extraneous-dependencies`. I'm not sure why that's not failing on the other libs used in the embeddings? Those are all marked as dev/peer dependencies as well.

File: langchain/package.json (hunk 1 above)
Comment (cwoolum) [reply langchain/package.json:679]: Yes, definitely. This went GA last week so it's very fresh.

Comment (cwoolum) [reply langchain/package.json:679]: I can take a look at updating the other integrations as well .

**Comment by cwoolum** [comment]
@jacoblee93 , fixed the linting issue. Let me know any other changes you'd like me to make. For docs, I tried to follow the existing pattern.

**Comment by jacoblee93** [comment]
Thank you so much for this!

I switched `modelName` -> `model` for consistency with the current Bedrock implementations, added some more initialization options in case the user doesn't want to pass a client directly, and added the entrypoint. Otherwise, looks great!

**Comment by cwoolum** [comment]
Sounds good! I'll work on porting the existing the other Bedrock integration to use the AWS SDK in the coming week.

**Comment by jacoblee93** [comment]
You rock!

File: langchain/package.json (hunk 1 above)
Comment (metaskills) [reply langchain/package.json:679]: Is this not yet done? If so, would you mind calling out those spots. Might be able to help with a PR too.

**Pull request merged**